File: go/idkollen/vipps.go
package idkollen

import (
	"context"
	"encoding/json"
	"fmt"
	"time"
)

// --- Vipps request types ---

type VippsAuthRequest struct {
	RedirectUrl    *string `json:"redirectUrl,omitempty"`
	RequestSsn     *bool   `json:"requestSsn,omitempty"`
	RequestPhone   *bool   `json:"requestPhone,omitempty"`
	RequestEmail   *bool   `json:"requestEmail,omitempty"`
	RequestAddress *bool   `json:"requestAddress,omitempty"`
	RefId          *string `json:"refId,omitempty"`
	AppCallbackUri *string `json:"appCallbackUri,omitempty"`
}

type VippsBackchannelAuthRequest struct {
	Phone          string  `json:"phone"`
	RequestSsn     *bool   `json:"requestSsn,omitempty"`
	RequestEmail   *bool   `json:"requestEmail,omitempty"`
	RequestAddress *bool   `json:"requestAddress,omitempty"`
	CallbackUrl    *string `json:"callbackUrl,omitempty"`
	RefId          *string `json:"refId,omitempty"`
}

// --- Vipps status types ---

type VippsStatus interface {
	isVippsStatus()
}

type VippsPending struct {
	Id    string  `json:"id"`
	RefId *string `json:"refId"`
	Url   *string `json:"url"`
}

type VippsCompleted struct {
	Id        string  `json:"id"`
	RefId     *string `json:"refId"`
	Ssn       string  `json:"ssn"`
	Name      string  `json:"name"`
	GivenName string  `json:"givenName"`
	Surname   string  `json:"surname"`
	Phone     *string `json:"phone"`
	Email     *string `json:"email"`
	Address   *string `json:"address"`
	BirthDate *string `json:"birthDate"`
	Pid       *string `json:"pid"`
	BankId    *string `json:"bankId"`
}

type VippsFailed struct {
	Id    string  `json:"id"`
	RefId *string `json:"refId"`
	Error string  `json:"error"`
}

func (*VippsPending) isVippsStatus()   {}
func (*VippsCompleted) isVippsStatus() {}
func (*VippsFailed) isVippsStatus()    {}

type vippsStatusEnvelope struct {
	Status string `json:"status"`
}

func unmarshalVippsStatus(data []byte) (VippsStatus, error) {
	var env vippsStatusEnvelope
	if err := json.Unmarshal(data, &env); err != nil {
		return nil, err
	}
	switch env.Status {
	case "PENDING":
		var v VippsPending
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	case "COMPLETED":
		var v VippsCompleted
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	case "FAILED":
		var v VippsFailed
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	default:
		return nil, fmt.Errorf("unknown vipps status: %s", env.Status)
	}
}

// --- Vipps endpoint ---

type VippsEndpoint struct {
	client *IdkollenClient
}

func (e *VippsEndpoint) Auth(ctx context.Context, req VippsAuthRequest) (VippsStatus, error) {
	var raw json.RawMessage
	if err := e.client.post(ctx, "/v3/vipps/auth", req, &raw); err != nil {
		return nil, err
	}
	return unmarshalVippsStatus(raw)
}

func (e *VippsEndpoint) BackchannelAuth(ctx context.Context, req VippsBackchannelAuthRequest) (VippsStatus, error) {
	var raw json.RawMessage
	if err := e.client.post(ctx, "/v3/vipps/backchannel/auth", req, &raw); err != nil {
		return nil, err
	}
	return unmarshalVippsStatus(raw)
}

func (e *VippsEndpoint) AuthStatus(ctx context.Context, id string) (VippsStatus, error) {
	var raw json.RawMessage
	if err := e.client.get(ctx, "/v3/vipps/auth/"+id, &raw); err != nil {
		return nil, err
	}
	return unmarshalVippsStatus(raw)
}

func (e *VippsEndpoint) CancelAuth(ctx context.Context, id string) error {
	return e.client.delete(ctx, "/v3/vipps/auth/"+id)
}

func (e *VippsEndpoint) WaitForAuth(ctx context.Context, id string, opts PollOptions) (VippsStatus, error) {
	interval := opts.Interval
	if interval == 0 {
		interval = 2 * time.Second
	}
	for {
		status, err := e.AuthStatus(ctx, id)
		if err != nil {
			return nil, &WaitError{Err: err}
		}
		if _, ok := status.(*VippsPending); !ok {
			return status, nil
		}
		select {
		case <-ctx.Done():
			return nil, &WaitError{Timeout: true}
		case <-time.After(interval):
		}
	}
}

File: ruby/lib/idkollen/vipps.rb
# frozen_string_literal: true

module Idkollen
  module Vipps
    class AuthRequest
      attr_reader :redirect_url, :request_ssn, :request_phone, :request_email,
                  :request_address, :ref_id, :app_callback_uri

      def initialize(redirect_url: nil, request_ssn: nil, request_phone: nil,
                     request_email: nil, request_address: nil, ref_id: nil,
                     app_callback_uri: nil)
        @redirect_url = redirect_url; @request_ssn = request_ssn
        @request_phone = request_phone; @request_email = request_email
        @request_address = request_address; @ref_id = ref_id
        @app_callback_uri = app_callback_uri
      end

      def to_h
        {
          redirectUrl: @redirect_url, requestSsn: @request_ssn,
          requestPhone: @request_phone, requestEmail: @request_email,
          requestAddress: @request_address, refId: @ref_id,
          appCallbackUri: @app_callback_uri,
        }.compact
      end
    end

    class BackchannelAuthRequest
      attr_reader :phone, :request_ssn, :request_email, :request_address,
                  :callback_url, :ref_id

      def initialize(phone:, request_ssn: nil, request_email: nil,
                     request_address: nil, callback_url: nil, ref_id: nil)
        @phone = phone; @request_ssn = request_ssn
        @request_email = request_email; @request_address = request_address
        @callback_url = callback_url; @ref_id = ref_id
      end

      def to_h
        {
          phone: @phone, requestSsn: @request_ssn,
          requestEmail: @request_email, requestAddress: @request_address,
          callbackUrl: @callback_url, refId: @ref_id,
        }.compact
      end
    end

    class Pending
      attr_reader :id, :ref_id, :url

      def initialize(id:, ref_id: nil, url: nil)
        @id = id; @ref_id = ref_id; @url = url
      end

      def deconstruct_keys(_) = { id: @id, ref_id: @ref_id, url: @url }

      def self.from_json(d) = new(id: d['id'], ref_id: d['refId'], url: d['url'])
    end

    class Completed
      attr_reader :id, :ref_id, :ssn, :name, :given_name, :surname,
                  :phone, :email, :address, :birth_date, :pid, :bank_id

      def initialize(id:, ssn:, name:, given_name:, surname:, ref_id: nil,
                     phone: nil, email: nil, address: nil, birth_date: nil,
                     pid: nil, bank_id: nil)
        @id = id; @ref_id = ref_id; @ssn = ssn; @name = name
        @given_name = given_name; @surname = surname; @phone = phone
        @email = email; @address = address; @birth_date = birth_date
        @pid = pid; @bank_id = bank_id
      end

      def deconstruct_keys(_)
        { id: @id, ref_id: @ref_id, ssn: @ssn, name: @name,
          given_name: @given_name, surname: @surname, phone: @phone,
          email: @email, address: @address, birth_date: @birth_date,
          pid: @pid, bank_id: @bank_id }
      end

      def self.from_json(d)
        new(
          id: d['id'], ref_id: d['refId'], ssn: d['ssn'], name: d['name'],
          given_name: d['givenName'], surname: d['surname'], phone: d['phone'],
          email: d['email'], address: d['address'], birth_date: d['birthDate'],
          pid: d['pid'], bank_id: d['bankId'],
        )
      end
    end

    class Failed
      attr_reader :id, :ref_id, :error

      def initialize(id:, error:, ref_id: nil)
        @id = id; @ref_id = ref_id; @error = error
      end

      def deconstruct_keys(_) = { id: @id, ref_id: @ref_id, error: @error }

      def self.from_json(d) = new(id: d['id'], ref_id: d['refId'], error: d['error'])
    end

    module Status
      def self.from_json(d)
        case d['status']
        when 'PENDING'   then Pending.from_json(d)
        when 'COMPLETED' then Completed.from_json(d)
        when 'FAILED'    then Failed.from_json(d)
        else raise ArgumentError, "Unknown vipps status: #{d['status']}"
        end
      end
    end

    class Endpoint
      def initialize(transport)
        @transport = transport
      end

      def auth(req)             = Status.from_json(@transport.post('/v3/vipps/auth', req.to_h))
      def backchannel_auth(req) = Status.from_json(@transport.post('/v3/vipps/backchannel/auth', req.to_h))
      def auth_status(id)       = Status.from_json(@transport.get("/v3/vipps/auth/#{id}"))
      def cancel_auth(id)       = @transport.delete("/v3/vipps/auth/#{id}")

      def wait_for_auth(id, opts: PollOptions.new) = poll(opts) { auth_status(id) }

      private

      def poll(opts)
        deadline = Time.now + opts.timeout
        loop do
          status = yield
          return status unless status.is_a?(Pending)
          raise WaitError.new(timeout: true) if Time.now >= deadline
          sleep opts.interval
        end
      end
    end
  end
end
